### PR TITLE
UI Outlined Icon

### PR DESCRIPTION
**Pull request by Amstutz**
While doing the final review on: #1194 (UI Card Extension) for the feature https://docu.ilias.de/goto_docu_wiki_wpage_4921_1357.html, I observed, the accepted version of the feature is asking for an outlined icons to be shown in this card. However, at the moment the card fails to provide that, since we do not yet have a proposal for the outlined icon yet.

Since we have the final JF on monday, I hereby add the proposal of how we can integrate the outlined icon in the KS for 5.4, so they can be used in the cards.

I decided to propose them as a variation version of the standard item, since to me, they are exactly that, Standard Items just rendered a little differently. 

The changes of the rules can be found [here](https://github.com/ILIAS-eLearning/ILIAS/compare/trunk...Amstutz:UI_Outlined_Icon?expand=1#diff-4d3ed3eabd7430039135d6f01b891b46).

Here is how they will look:
image

### CODE REVIEW

File: src/UI/Component/Icon/Factory.php [hunk 1]
```diff
@@ -9,14 +9,30 @@
 	 * ---
 	 * description:
 	 *   purpose: >
-	 *    Standard Icons represent ILIAS Objects.
-	 *   composition: >
+	 *    Standard Icons represent ILIAS Objects. In their outlined version, they are used to be layered upon a picture or
+     *    if the Standard Icon should not too much of the users attention.
+     *   composition: >
```
Comment (Amstutz) [inline src/UI/Component/Icon/Factory.php]: sry, something seems of here with the intends. however, the yaml format seems intact. Ill correct that.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 15 OCT 2018 : We highly appreciate this suggestion and schedule the PR for 5.4. It would be great to use the outline icons already for the deck of card implementation for 5.4. Thanks a lot to Yvonne for providing all outlined icons!

**Comment by yvseiler** [comment]
> Jour Fixe, 15 OCT 2018 : We highly appreciate this suggestion and schedule the PR for 5.4. It would be great to use the outline icons already for the deck of card implementation for 5.4. Thanks a lot to Yvonne for providing all outlined icons!

This is great and will be a pleasure to implement. A big thank you goes to CaT, namely Philipp Geib and Christoph Krahl, who have already worked actively on the implementation of the first outlined icons and have great influence on the design guidelines for the outlined icon proposals.

We are putting the pedal to the metal.

**Comment by xus** [comment]
Is this ready to be merged? I need to use this on the new UI Cards ASAP and merge all before the release.
https://github.com/ILIAS-eLearning/ILIAS/pull/1194

**Comment by Amstutz** [comment]
Hi @xus: I left a comment in the linke PR (see #1194 ). The outlined Icons have been accepted. However, Yvonne needs some time to finish up the final set in collaboration with CaT. Further I do not like to merge my own PR's. So I will wait with that until Richard is back an gives his final review. If everything else is resolved I will merge this without the outlined Icons yet available so you can proceed.

File: src/UI/Component/Icon/Factory.php [hunk 2]
```diff
@@ -9,13 +9,29 @@
 	 * ---
 	 * description:
 	 *   purpose: >
-	 *    Standard Icons represent ILIAS Objects.
+	 *    Standard Icons represent ILIAS Objects. In their outlined version, they are used to be layered upon a picture or
+	 *    if the Standard Icon should not too much of the users attention.
```
Comment (klees) [inline src/UI/Component/Icon/Factory.php]: I think a `catch`or something similar is missing here:

```suggestion
	 *    if the Standard Icon should not catch too much of the users attention.
```

File: src/UI/Component/Icon/Factory.php [hunk 3]
```diff
@@ -9,13 +9,29 @@
...
 	 *   composition: >
 	 *     A Standard Icon is displayed as a block-element with a background-graphic.
 	 *     By default, a fallback icon will be rendered; this is until a
-	 *     background image is defined in the icon's CSS-class.
+	 *     background image is defined in the icon's CSS-class. The outlined version is the silhouette of the Standard Icon.
+	 *     This outlined version attracts therefore less attention.
```
Comment (klees) [inline src/UI/Component/Icon/Factory.php]:
```suggestion
	 *     This outlined version therefore attracts less attention.
```

File: src/UI/Component/Icon/Factory.php [hunk 4]
```diff
@@ -9,13 +9,29 @@
...
+	 * rules:
+	 *   usage:
+	 *     1: If they are layered upon a picture, the outlined version MUST be used.
+	 *     2:  >
+	 *        In any other case, the non-outlined version SHOULD be used except for scenarios where the icon should
+	 *        not draw too much attention.
+	 *   style:
+	 *     1: CSS-Filters MAY be used for Standard Icons in their outlined version to manipulate the stroke to fit the context.
+	 *     2: >
+	 *        In their outlined version, Standard Icons MUST only white as color for the stroke, to make filter easily
```
Comment (klees) [inline src/UI/Component/Icon/Factory.php]:
```suggestion
	 *        In their outlined version, Standard Icons MUST only use white as color for the stroke, to make filter easily
```

File: src/UI/Implementation/Component/Icon/Standard.php [hunk 5]
```diff
@@ -138,4 +144,22 @@ public function getAllStandardHandles() {
...
+    /**
+     * @param bool $is_outlined
+     * @return Icon
+     */
+    public function withIsOutlined(bool $is_outlined)
+    {
+        $clone = clone $this;
+        $clone->is_outlined = $is_outlined;
+        return $clone;
+    }
```
Comment (klees) [inline src/UI/Implementation/Component/Icon/Standard.php]: I think we are mixing indentations here. The spaces match our general decision to use PSR-2 some times in the future, but I think we should not mix in one file.

File: src/UI/Implementation/Component/Icon/Standard.php [hunk 6]
```diff
@@ -138,4 +144,22 @@ public function getAllStandardHandles() {
 		return self::$standard_icons;
 	}
 
+    /**
+     * @return bool
+     */
+    public function isOutlined(): bool
```
Comment (klees) [inline src/UI/Implementation/Component/Icon/Standard.php]: :+1: for return-type-hint.

File: src/UI/Implementation/Component/Icon/Standard.php [hunk 7]
```diff
@@ -138,4 +144,22 @@ public function getAllStandardHandles() {
...
+    {
+        return $this->is_outlined;
+    }
+
+    /**
+     * @param bool $is_outlined
+     * @return Icon
+     */
+    public function withIsOutlined(bool $is_outlined)
```
Comment (klees) [inline src/UI/Implementation/Component/Icon/Standard.php]: Why no return-type hint here?

File: src/UI/examples/Icon/Standard/disabled_icon.php [hunk 8]
```diff
@@ -7,17 +7,11 @@ function disabled_icon()
...
+	$ico = $f->icon()->standard('grp', 'Course', 'large', false)->withDisabled(true);
 
-	$buffer[] = $renderer->render($ico->withDisabled(true))
-		.' Small Course';
-
-	$buffer[] = $renderer->render($ico->withSize('medium')->withDisabled(true))
-		.' Medium Course';
-
-	$buffer[] = $renderer->render($ico->withSize('large')->withDisabled(true))
-		.' Large Course';
```
Comment (klees) [inline src/UI/examples/Icon/Standard/disabled_icon.php:19]: Why did you remove the other sizes?

**Review by klees — CHANGES_REQUESTED** [review]
LGTM in general. I do have some minor editorial changes and some questions though.

File: src/UI/examples/Icon/Standard/disabled_icon.php (hunk 8 above)
Comment (Amstutz) [reply src/UI/examples/Icon/Standard/disabled_icon.php:19]: Because this example is about disabling, not about sizes. We have another one for that. To me it is more important to show disabling for outlined and non-outlined than various sizes. If we do both, we have an overload here.

**Comment by Amstutz** [comment]
Hi Richard, thx a lot for the feedback. I hope to have answered all your questions, and fixed any open pending issues. If so, feel free to hit merge, if not let me know.

Note, that we provided completed set of outlined icons, therefore the many files marked as changed.

**Comment by klees** [comment]
@Amstutz I think you did not update the `data.json` for the documentation, I was missing the mandatory conflict in that file. I expect that file to be updated some time in the next days, so I pushed the button anyway.
